src/docfx/lib/html/HtmlReader.cs

```diff
@@ -0,0 +1,443 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Buffers;
+
+namespace Microsoft.Docs.Build
+{
+    /// <summary>
+    /// Provides a high-performance, near-zero-allocation, fault-tolerant, standard-compliant API
+    /// for forward-only, read-only access to HTML text.
+    /// </summary>
+    internal ref struct HtmlReader
+    {
+        private string _html;
+        private int _length;
+        private int _position;
+
+        private HtmlTokenType _type;
+        private (int start, int length) _nameRange;
+        private (int start, int length) _tokenRange;
+
+        private (int start, int length) _attributeNameRange;
+        private (int start, int length) _attributeValueRange;
+        private (int start, int length) _attributeRange;
+        private HtmlAttribute[]? _attributes;
+        private int _attributesLength;
+
+        public HtmlTokenType Type => _type;
+
+        public (int start, int length) NameRange => _nameRange;
+
+        public (int start, int length) ContentRange => _tokenRange;
+
+        public ReadOnlySpan<char> Name => _html.AsSpan().Slice(_nameRange.start, _nameRange.length);
```

Comment:
Can I think of `AsSpan()` is like creating a pointer to specific block of memory(here the string's still allocated on heap?)

If so, can `AsSpan()` be called only once and cached, instead of many time?

Reply:
https://docs.microsoft.com/en-us/archive/msdn-magazine/2018/january/csharp-all-about-span-exploring-a-new-net-mainstay